spec.bs

```diff
@@ -1134,6 +1006,197 @@ To <dfn>convert a modifier to a string</dfn> given a [=part/modifier=] |modifier
+  1. If |init|["{{URLPatternInit/pathname}}"] is not null then set |result|["{{URLPatternInit/pathname}}"] to the result of [=process pathname for init=] given |init|["{{URLPatternInit/pathname}}"], |result|["{{URLPatternInit/protocol}}"], and |type|.
+  1. If |init|["{{URLPatternInit/search}}"] is not null then set |result|["{{URLPatternInit/search}}"] to the result of [=process search for init=] given |init|["{{URLPatternInit/search}}"] and |type|.
+  1. If |init|["{{URLPatternInit/hash}}"] is not null then set |result|["{{URLPatternInit/hash}}"] to the result of [=process hash for init=] given |init|["{{URLPatternInit/hash}}"] and |type|.
+  1. Return |result|.
+</div>
+
+<div algorithm>
+  To <dfn>process protocol for init</dfn> given a string |value| and a string |type|:
+
+  1. Let |strippedValue| be the given |value| with a single trailing U+003A (`:`) removed, if any.
+  1. If |type| is "`pattern`" then return |strippedValue|.
+  1. Return the result of running [=canonicalize a protocol=] given |strippedValue|.
+</div algorithm>
+
+<div algorithm>
+  To <dfn>process username for init</dfn> given a string |value| and a string |type|:
+
+  1. If |type| is "`pattern`" then return |value|.
+  1. Return the result of running [=canonicalize a username=] given |value|.
+</div algorithm>
+
+<div algorithm>
+  To <dfn>process password for init</dfn> given a string |value| and a string |type|:
+
+  1. If |type| is "`pattern`" then return |value|.
+  1. Return the result of running [=canonicalize a password=] given |value|.
+</div algorithm>
+
+<div algorithm>
+  To <dfn>process hostname for init</dfn> given a string |value| and a string |type|:
+
+  1. If |type| is "`pattern`" then return |value|.
+  1. Return the result of running [=canonicalize a hostname=] given |value|.
+</div algorithm>
+
+<div algorithm>
+  To <dfn>process port for init</dfn> given a string |portValue|, a string |protocolValue|, and a string |type|:
+
+  1. If |type| is "`pattern`" then return |portValue|.
+  1. Return the result of running [=canonicalize a port=] given |portValue| and |protocolValue|.
+</div algorithm>
+
+<div algorithm>
+  To <dfn>process pathname for init</dfn> given a string |pathnameValue|, a string |protocolValue|, and a string |type|:
+
+  1. If |type| is "`pattern`" then return |pathnameValue|.
+  1. Let |dummyURL| be a new [=URL record=].
+  1. If |protocolValue| is null or the empty string, then set |protocolValue| to "`https`".
+  1. Set |dummyURL|'s [=url/scheme=] to |protocolValue|.
+     <p class="note">Note, we set the [=URL record=]'s [=url/scheme=] in order to check if it [=is special=].  If |protocolValue| is missing we purposely default the URL to being special by using "`https`" for the [=url/scheme=].</p>
+  1. If |dummyURL| [=is special=], then:
+    1. Return the result of running [=canonicalize a standard pathname=] given |pathnameValue|.
```

Comment:
Nit: I personally prefer inlining single-line blocks into their "if" steps, i.e.

1. If dummyURL is special, then return the result of...

Up to you though.